R/utils.R

```diff
@@ -25,8 +25,19 @@
 #'  with text placed before the output to put the output into context. For example a title.
 #' @param post_output (`shiny.tag`) optional, text or UI element to be displayed after the module's output,
 #' adding context or further instructions. Elements like `shiny::helpText()` are useful.
-#' @param decorators `r lifecycle::badge("experimental")` (`list` of `teal_transform_module` or `NULL`) optional,
+#' @param decorators `r lifecycle::badge("experimental")`
+#' (`list` of `teal_transform_module`,
+#' named `list` of `teal_transform_module` or
+#' `NULL`) optional,
 #' if not `NULL`, decorator for tables or plots included in the module.
+#' When a named list of `teal_transform_module`, the decorators are applied to the
+#' outputs of the corresponding view of the module.
+#' Otherwise, the decorators are applied to all outputs in all views equally, which is the same
+#' as using the name `default`.
+#'
+#' A module can have different views with different outputs, such as a tab with
+#' a table and another with a plot.
+#' See section "Decorating `tm_<module_name>`" below for more details.
```

Comment:
@m7pr @llrs-roche I'd appreciate opinion/suggestion here as well :-)

It seems confusing.

Reply:
@averissimo if you pay attention to this comment https://github.com/insightsengineering/teal.modules.general/pull/812#discussion_r1860533448
then I think we don't confuse people with `outputs of the corresponding view of the module` and actual internal object names.

Let's just use internal object names. This is a very advanced feature, just for people who will need to look inside the code anyway. So including UI names in here is too much

Reply:
Yes, with this feature we expose some internal objects to other developers, I too would keep the internal object names even if using the more descriptive name would be more meaningful.

Reply:
I guess there are 2 competing approaches here:

## Scenario

Let's say he have 2 tabs, `histogram` & `qq`

With independent `qenvs`

- `hist_q`: uses `plot_1` and `table`
- `qq_q`: uses `table` and `plot_2`

### Object-based approach

Let's say he have 2 tabs, `histogram` & `qq`

With independent `qenvs`

- `hist_q`: uses `plot_1` and `table`
- `qq_q`: uses `table` and `plot_2`

The code for `ui_decorate_teal_data` and `srv_decorate_teal_data` would be:

```diff
# call to module
tm_my_module(
  decorators = list(
    default = list(dec_all),
+    plot_1 = list(dec_plot1),
+    plot_2 = list(dec_plot),
+    table  = list(dec_table)
  )
)
```

```r
# ui/srv pair for histogram
ui_decorate_teal_data(ns("d_hist"), decorators = subset_decorators(c("plot_1", "table"), args$decorators)),
srv_decorate_teal_data("d_hist", decorators = subset_decorators(c("plot_1", "table"), decorators)),

# ui/srv pair for qq
ui_decorate_teal_data(ns("d_qq"), decorators = subset_decorators(c("table", "plot_2"), args$decorators)),
srv_decorate_teal_data("d_qq", decorators = subset_decorators(c("table", "plot_2"), decorators))
```

## View-based approach

The code for `ui_decorate_teal_data` and `srv_decorate_teal_data` would be:

```diff
# call to module
tm_my_module(
  decorators = list(
    default = list(dec_all),
+    histogram = list(dec_plot1), # note that `dec_table` is not applied
+    qq = list(dec_table, dec_plot_2)
  )
)
```

```r
# ui/srv pair for histogram
ui_decorate_teal_data(ns("d_hist"), decorators = subset_decorators(c("histogram"), args$decorators)),
srv_decorate_teal_data("d_hist", decorators = subset_decorators(c("histogram"), decorators)),

# ui/srv pair for qq
ui_decorate_teal_data(ns("d_qq"), decorators = subset_decorators(c("qq"), args$decorators)),
srv_decorate_teal_data("d_qq", decorators = subset_decorators(c("qq"), decorators))
```

Reply:
I'm kinda sold at the `object-based` approach

### Object-based approach 

- Simple documentation (👍👍👍)
- Easier to understand
- Keeps low level

### View-base approach

- Allows for "per view" decorators (👍)
- Higher-level approach